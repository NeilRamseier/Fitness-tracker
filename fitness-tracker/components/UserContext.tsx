// UserContext.tsx

import React, { createContext, useContext, useState, ReactNode } from 'react';

interface User {
  id?: number;
  first_name?: string;
  last_name?: string;
  weight?: number;
  height?: number;
  age?: number;
  gender?: string;
  basal_metabolic_rate?: number;
  IP?: string;
  creation_date?: string;
}

interface UserContextType {
  user: User;
  setUser: React.Dispatch<React.SetStateAction<User>>;
  calories: number | null;
  setCalories: React.Dispatch<React.SetStateAction<number | null>>;
}

const UserContext = createContext<UserContextType | undefined>(undefined);

export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<User>({});
  const [calories, setCalories] = useState<number | null>(null);

  return (
    <UserContext.Provider value={{ user, setUser, calories, setCalories }}>
      {children}
    </UserContext.Provider>
  );
};

export const useUser = (): UserContextType => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }
  return context;
};
