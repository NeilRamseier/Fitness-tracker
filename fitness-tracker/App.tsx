import * as React from "react";
import { Text, View } from "react-native";
import { NavigationContainer } from "@react-navigation/native";
import { createBottomTabNavigator } from "@react-navigation/bottom-tabs";
import { DefaultTheme, Icon, PaperProvider } from "react-native-paper";
import HomeScreen from "./components/HomeScreen";
import KalenderScreen from "./components/KalenderScreen";
import ProfilScreen from "./components/ProfilScreen";
import MaterialCommunityIcons from "react-native-vector-icons/MaterialCommunityIcons";
import SportsUnitScreen from "./components/SportsUnitScreen";

const Tab = createBottomTabNavigator();

const theme = {
  ...DefaultTheme,
  colors: {
    ...DefaultTheme.colors,
    primary: "#1A5A41",
    secondary: "#EAE7E7",
  },
};

export default function App() {
  return (
    <PaperProvider theme={theme}>
      <NavigationContainer>
        <Tab.Navigator
          screenOptions={{
            tabBarShowLabel: false,
          }}
        >
          <Tab.Screen
            name="Home"
            component={HomeScreen}
            options={{
              tabBarStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerTitleStyle: {
                color: theme.colors.secondary,
              },
              tabBarIcon: () => (
                <MaterialCommunityIcons
                  name="home"
                  color={theme.colors.secondary}
                  size={28}
                />
              ),
            }}
          />
          <Tab.Screen
            name="SportsUnit"
            component={SportsUnitScreen}
            options={{
              tabBarStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerTitleStyle: {
                color: theme.colors.secondary,
              },
              title: "Sporteinheit",
              tabBarIcon: ({ color, size }) => (
                <MaterialCommunityIcons
                  name="plus-circle"
                  color={theme.colors.secondary}
                  size={28}
                />
              ),
            }}
          />
          <Tab.Screen
            name="Kalender"
            component={KalenderScreen}
            options={{
              tabBarStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerTitleStyle: {
                color: theme.colors.secondary,
              },
              tabBarIcon: ({ color, size }) => (
                <MaterialCommunityIcons
                  name="calendar"
                  color={theme.colors.secondary}
                  size={28}
                />
              ),
            }}
          />
          <Tab.Screen
            name="Profil"
            component={ProfilScreen}
            options={{
              tabBarStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerStyle: {
                backgroundColor: theme.colors.primary,
              },
              headerTitleStyle: {
                color: theme.colors.secondary,
              },
              tabBarIcon: ({ color, size }) => (
                <MaterialCommunityIcons
                  name="account"
                  color={theme.colors.secondary}
                  size={28}
                />
              ),
            }}
          />
        </Tab.Navigator>
      </NavigationContainer>
    </PaperProvider>
  );
}
