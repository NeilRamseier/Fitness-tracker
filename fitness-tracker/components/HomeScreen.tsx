import { View, Text, StyleSheet, TouchableOpacity } from "react-native";
import * as React from 'react';
import { useTheme, Surface, Icon } from "react-native-paper";
import { Pedometer } from "expo-sensors";
import { changeUserWeight, currentUser } from "../App";
import { useUser } from "./UserContext";

export default function HomeScreen() {
  const theme = useTheme();
  const { user, setUser, basal  } = useUser();
  const [counter, setCounter] = React.useState<number | null>(null); // null, bis Daten geladen sind
  const [isPedometerAvailable, setIsPedometerAvailable] = React.useState('checking');
  const [stepCount, setStepCount] = React.useState(0);
  const { calories } = useUser();
  const [localBasal, setLocalBasal] = React.useState<number | null>(basal);

  const MIN_COUNT = 0;


  React.useEffect(() => {
    if (user && user.weight) {
      setCounter(user.weight); // Initialisiere Gewicht aus user
    } else {
      setCounter(0.1); // Standardwert, falls kein Benutzer vorhanden
    }
  }, [user]); // Aktiviere, wenn sich currentUser ändert

  // UseEffect to update basal whenever user.basal_metabolic_rate changes
  React.useEffect(() => {
    if (user && user.basal_metabolic_rate) {
      setLocalBasal(user.basal_metabolic_rate);  // Update local basal state
    }
  }, [user]);

  // Pedometermessung initialisieren
  const subscribe = async () => {
    const isAvailable = await Pedometer.isAvailableAsync();
    setIsPedometerAvailable(String(isAvailable));

    if (isAvailable) {
      const today = new Date();
      const start = new Date(today);
      start.setHours(0, 0, 0, 0);
      const end = new Date(today);
      end.setHours(23, 59, 59, 999);

      const stepCountResult = await Pedometer.getStepCountAsync(start, end);
      setStepCount(stepCountResult.steps);
    }
  };

  React.useEffect(() => {
    const subscription = subscribe();
  }, []);

  // Erhöhe das Gewicht
  const higherCounter = () => {
    if (counter !== null) {
      const newWeight = Math.max(Math.round((counter + 0.1) * 10) / 10, 0.1);
      setCounter(newWeight);
      changeUserWeight(newWeight);
      if (user) {
        setUser({ ...user, weight: newWeight }); 
      }
    }
  };

  // Verringere das Gewicht
  const lowerCounter = () => {
    if (counter !== null) {
      const newWeight = Math.max(Math.round((counter - 0.1) * 10) / 10, 0.1);
      setCounter(newWeight);
      changeUserWeight(newWeight);
      if (user) {
        setUser({ ...user, weight: newWeight }); // Update user context
      }
    }
  };

  return (
    <View style={{ backgroundColor: theme.colors.primary, flex: 1, justifyContent: "flex-start", alignItems: "center" }}>
      <Text style={{ color: theme.colors.secondary, fontSize: 35, marginTop: 50 }}>Hallo User!</Text>
      
      {/* Schritte anzeigen */}
      <Surface
        style={{
          justifyContent: "center",
          alignItems: "center",
          marginTop: 50,
          backgroundColor: theme.colors.primary,
          borderColor: theme.colors.secondary,
          width: 350,
          height: 80,
          borderWidth: 1,
          borderRadius: 10
        }}
        elevation={3}
      >
        <Text
          style={{
            fontSize: 25,
            color: theme.colors.secondary
          }}
        >
          {stepCount} Schritte
        </Text>
      </Surface>

      {/* Kalorienverbrauch anzeigen */}
      <Surface
        style={{
          display: "flex",
          flexDirection: "row",
          justifyContent: "space-around",
          alignItems: "center",
          marginTop: 25,
          marginBottom: 25,
          backgroundColor: theme.colors.primary,
          borderColor: theme.colors.secondary,
          width: 350,
          height: 150,
          borderWidth: 1,
          borderRadius: 10
        }}
        elevation={3}
      >
        <Text
          style={{
            fontSize: 25,
            color: theme.colors.secondary
          }}
        >
           {localBasal !== null ? `${Math.round((calories ?? 0) + (localBasal ?? 0))} kcal` : "0"}
        </Text>
        {/* Detailverbrauch */}
        <View>
          <Text style={styles.detailText}>Grundverbrauch</Text>
          <Text style={styles.detailText}>{localBasal !== null ? localBasal : "0"} kcal</Text>
          <Text style={styles.detailText}>Aktivitäten</Text>
          <Text style={styles.detailText}>{calories !== null ? `${calories.toFixed(2)} kcal` : "0"}</Text>
        </View>
      </Surface>

      {/* Gewichtsübersicht */}
      <TouchableOpacity onPress={higherCounter} testID="increase-button">
        <View style={[styles.plusAndMinus, { borderColor: theme.colors.secondary }]}>
          <Icon source="plus" color={theme.colors.secondary} size={50} />
        </View>
      </TouchableOpacity>

      {/* Anzeige des Benutzergewichts */}
      <View style={{ flexDirection: "row", alignItems: "flex-end" }}>
        <Text style={{ color: theme.colors.secondary, fontSize: 50, marginBottom: 5 }}>
        {user?.weight !== undefined ? `${user.weight} kg` : "Nicht verfügbar"}
        </Text>
        <Text style={{ color: theme.colors.secondary, fontSize: 15, marginBottom: 5, marginLeft: 5 }}>Kg</Text>
      </View>
      <View style={{ backgroundColor: theme.colors.secondary, height: 1, width: 180, marginBottom: 20 }} />

      <TouchableOpacity onPress={lowerCounter} testID="decrease-button">
        <View style={[styles.plusAndMinus, { borderColor: theme.colors.secondary }]}>
          <Icon source="minus" color={theme.colors.secondary} size={50} />
        </View>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  detailText: {
    marginTop: 5,
    fontSize: 20,
    color: "#808080"
  },
  plusAndMinus: {
    width: 60,
    height: 60,
    marginBottom: 10,
    borderWidth: 3,
    borderRadius: 50,
    justifyContent: "center",
    alignItems: "center"
  }
});
