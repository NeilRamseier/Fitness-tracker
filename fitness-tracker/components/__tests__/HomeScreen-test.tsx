import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import HomeScreen from '../HomeScreen';
import { UserProvider, useUser } from '../UserContext';
import { changeUserWeight } from '../../App'; // Make sure to import the mocked function

// Mocking the `changeUserWeight` function
jest.mock('../../App', () => ({
  changeUserWeight: jest.fn(),
}));

// Mock the `useUser` hook
jest.mock('../UserContext', () => {
  const actual = jest.requireActual('../UserContext');
  return {
    ...actual,
    useUser: jest.fn(),
  };
});

describe('HomeScreen', () => {
  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
  });

  it('should increase weight when plus button is pressed', () => {
    const mockSetUser = jest.fn();
    const mockUser = { weight: 70, basal_metabolic_rate: 1500 };

    (useUser as jest.Mock).mockReturnValue({
      user: mockUser,
      setUser: mockSetUser,
      calories: 200,
      setCalories: jest.fn(),
      basal: 1500,
      setBasal: jest.fn(),
    });

    const { getByTestId } = render(
      <UserProvider>
        <HomeScreen />
      </UserProvider>
    );

    const increaseButton = getByTestId('increase-button');
    
    fireEvent.press(increaseButton);
    
    expect(mockSetUser).toHaveBeenCalledWith({ ...mockUser, weight: 70.1 });
  });

  it('should not decrease weight below 0.1 kg', () => {
    const mockSetUser = jest.fn();
    const mockUser = { weight: 0.1, basal_metabolic_rate: 1500 };

    (useUser as jest.Mock).mockReturnValue({
      user: mockUser,
      setUser: mockSetUser,
      calories: 200,
      setCalories: jest.fn(),
      basal: 1500,
      setBasal: jest.fn(),
    });

    const { getByTestId } = render(
      <UserProvider>
        <HomeScreen />
      </UserProvider>
    );

    const decreaseButton = getByTestId('decrease-button');
    
    fireEvent.press(decreaseButton);
    
    expect(mockSetUser).not.toHaveBeenCalledWith({ ...mockUser, weight: 0.0 });
  });
});


describe('<HomeScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <UserProvider> {/* Wrappen des HomeScreen mit UserProvider */}
          <HomeScreen />
        </UserProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});