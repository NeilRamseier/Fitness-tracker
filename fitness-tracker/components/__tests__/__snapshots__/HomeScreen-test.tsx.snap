// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> renders correctly 1`] = `
[
  " ",
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(103, 80, 164, 1)",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      style={
        {
          "color": "rgba(98, 91, 113, 1)",
          "fontSize": 35,
          "marginTop": 50,
        }
      }
    >
      Hallo User!
    </Text>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(103, 80, 164, 1)",
          "borderRadius": 10,
          "height": 80,
          "marginTop": 50,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 8,
          "width": 350,
        }
      }
      testID="surface-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(103, 80, 164, 1)",
            "borderColor": "rgba(98, 91, 113, 1)",
            "borderRadius": 10,
            "borderWidth": 1,
            "flex": 1,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 3,
          }
        }
        testID="surface"
      >
        <Text
          style={
            {
              "color": "rgba(98, 91, 113, 1)",
              "fontSize": 25,
            }
          }
        >
          0
           Schritte
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(103, 80, 164, 1)",
          "borderRadius": 10,
          "height": 150,
          "marginBottom": 25,
          "marginTop": 25,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 8,
          "width": 350,
        }
      }
      testID="surface-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(103, 80, 164, 1)",
            "borderColor": "rgba(98, 91, 113, 1)",
            "borderRadius": 10,
            "borderWidth": 1,
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 3,
          }
        }
        testID="surface"
      >
        <Text
          style={
            {
              "color": "rgba(98, 91, 113, 1)",
              "fontSize": 25,
            }
          }
        >
          1700 kcal
        </Text>
        <View>
          <Text
            style={
              {
                "color": "#808080",
                "fontSize": 20,
                "marginTop": 5,
              }
            }
          >
            Grundverbrauch
          </Text>
          <Text
            style={
              {
                "color": "#808080",
                "fontSize": 20,
                "marginTop": 5,
              }
            }
          >
            1500
             kcal
          </Text>
          <Text
            style={
              {
                "color": "#808080",
                "fontSize": 20,
                "marginTop": 5,
              }
            }
          >
            Aktivitäten
          </Text>
          <Text
            style={
              {
                "color": "#808080",
                "fontSize": 20,
                "marginTop": 5,
              }
            }
          >
            200.00 kcal
          </Text>
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="increase-button"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 50,
              "borderWidth": 3,
              "height": 60,
              "justifyContent": "center",
              "marginBottom": 10,
              "width": 60,
            },
            {
              "borderColor": "rgba(98, 91, 113, 1)",
            },
          ]
        }
      >
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            [
              {
                "color": "rgba(98, 91, 113, 1)",
                "fontSize": 50,
              },
              [
                {
                  "lineHeight": 50,
                  "transform": [
                    {
                      "scaleX": 1,
                    },
                  ],
                },
                {
                  "backgroundColor": "transparent",
                },
              ],
              {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰐕
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "flex-end",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "color": "rgba(98, 91, 113, 1)",
            "fontSize": 50,
            "marginBottom": 5,
          }
        }
      >
        0.1 kg
      </Text>
      <Text
        style={
          {
            "color": "rgba(98, 91, 113, 1)",
            "fontSize": 15,
            "marginBottom": 5,
            "marginLeft": 5,
          }
        }
      >
        Kg
      </Text>
    </View>
    <View
      style={
        {
          "backgroundColor": "rgba(98, 91, 113, 1)",
          "height": 1,
          "marginBottom": 20,
          "width": 180,
        }
      }
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="decrease-button"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 50,
              "borderWidth": 3,
              "height": 60,
              "justifyContent": "center",
              "marginBottom": 10,
              "width": 60,
            },
            {
              "borderColor": "rgba(98, 91, 113, 1)",
            },
          ]
        }
      >
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            [
              {
                "color": "rgba(98, 91, 113, 1)",
                "fontSize": 50,
              },
              [
                {
                  "lineHeight": 50,
                  "transform": [
                    {
                      "scaleX": 1,
                    },
                  ],
                },
                {
                  "backgroundColor": "transparent",
                },
              ],
              {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰍴
        </Text>
      </View>
    </View>
  </View>,
]
`;
